### PR TITLE
PitchShiftEffect: decrease and report latency

### PR DESCRIPTION
**Pull request by davidchocholaty**
This PR improves the current implementation of the Pitch shift effect. The implementation approach is still the "push" model as known. The new version decreases the amount of latency that the effect produces. The implementation is based on the public ladspa-lv2 example implementation by RubberBand. The main difference is, that the maximum number of samples, that will be provided into RubberBand is not set, but instead of it, the minimum required number of samples is provided and processed. Based on the processing of smaller chunks of the input buffer the latency is recognizably decreased.

To avoid the dropouts which are possible in the previous implementation, the specific amount of zero samples is prefilled into the RubberBand. Based on that, the sample dropouts may occur only for the extreme settings changes. The implementation is possible without prefilling the zero samples and the latency decreased even more, anyway, a greater amount of sample dropouts occur.

The known issue is, that the sound quality for the new implementation, primarily for the +12 semitones settings looks a little bit worse to me, but IMO it could be probably solved by the RubberBand option settings.

In this PR the latency is propagated into the delay handling of the effect chain too. For now, the initial delay based on the prefilled samples is provided. Based on that, the latency reporting works properly for the initial setting, however, after dynamic pitch changes, the propagated amount of delay is not valid. This issue should be fixed in this PR too.

Please share your thoughts, comments and proposed changes to the new implementation. Thank you.

### CODE REVIEW

**Comment by Swiftb0y** [comment]
I haven't looked to deeply into the changes, but if IIUC this PR basically caches the incoming and outgoing samples so it can provide rubberband whoever much it needs instead of however much it wants? I struggle to understand how that can reduce latency. Why doesn't rubberband do this caching automatically in the pull mode?

**Comment by davidchocholaty** [comment]
The main difference is, that in the previous version the maximum amount of frames that can be provided into the RubberBand is set. From some RubberBand discussion, it just serves to avoid inner overflows. However, when the maximum amount of frames is not sufficient for RubberBand, it takes after the effect turned on a specific amount of whole empty output buffers. Then some output frames are provided from RubberBand, but the number of frames is not enough to fill the whole output buffer. Based on that, some meaningful buffer is after a few buffers (empty and then not completely full).

For the new version, when the number of zero frames is filled into the RubberBand, it takes exactly the same amount of zero output samples, but the amount is not such big as in the previous version, so the delay is smaller. For the new version without prefilling the RubberBand, which was mentioned in the first message, the situation is similar, but compared to the previous version, it looks, that the amount of free or not full buffers are smaller too, if I remember it right.

In the previous version after the Mixxx app is run and the effect is firstly turned on when the track plays, you can actually hear the mentioned situation, when the starts buffer is not completely filled.

The implementation is not exactly the "pull" model as you wrote, for example from the following reasons:

1. Prefilling with **zero** samples to avoid dropouts.
2. When the last block of the input buffer is processed and the amount of frames that RubberBand requires is greater than the rest of the input buffer frames, **only the rest** of the input buffer frames are provided and not the amount, that RubberBand requires.

In summary, IMHO it looks that this approach suits the RubberBand more for real-time use.

**Comment by Swiftb0y** [comment]
I'm sorry but I still don't get it completely.

> The implementation is not exactly the "pull" model as you wrote

Just to clarify, we still reversed the interaction with rubberband. Rubberband is asking us for data instead of us handing it data. Correct?

> Prefilling with zero samples to avoid dropouts.

This is ambiguous to me. Do you mean we fill rubberband with a bunch of samples that are just 0 (so silence in practice), or do we call its process method but handing no samples at all (0 amount)? I assume we do the first one as that makes more sense. What kind of dropouts is that strategy supposed to prevent? Why do these dropouts occur?


So this version basically feeds rubberband silence until we have real audio data to serve? 


> When the last block of the input buffer is processed and the amount of frames that RubberBand requires is greater than the rest of the input buffer frames, only the rest of the input buffer frames are provided and not the amount, that RubberBand requires.

What advantage does this have?
---
And my main question is still, why does rubberband not do all of this internally by default if its better?

**Comment by davidchocholaty** [comment]
> Just to clarify, we still reversed the interaction with rubberband. Rubberband is asking us for data instead of us handing it data. Correct?

Yeah, that's right. To just clear it up a little bit, I think that we can call it a "hybrid" implementation between the "push" and the "pull" model as known. As you wrote, in the new version the RubberBand is asking us for the specific number of frames it needs. It is similar to the "pull" model, but it will be really "clean pull" model, it has to follow the specification from RubberBand's documentation:

_the number of samples you supply to it for a given number of output samples is allowed to vary freely. That is:_
-  _you have some sort of time-processing object, which holds a Rubber Band stretcher object;_
-  _your audio output driver or graph requests a certain number of samples from it_

However, what the new version only does is, that it splits the input buffer into smaller chunks if RubberBand does not require so much data for processing. For example for setting +12 semitones with 1024 frames per input buffer, after some processed buffers the RubberBand requires only about 264 frames (`the information from getSamplesRequired()`) for one `process()` call, so the buffer is split into four parts:

the input buffer 1024 frames:
- 264 frames
- 264 frames
- 264 frames
- **234** frames

For exactly the "clean pull" model implementation, the last value has to be 264 frames too, because RubberBand requires it, but we can only offer the rest frames of the input buffer.

**Comment by davidchocholaty** [comment]
> This is ambiguous to me. Do you mean we fill rubberband with a bunch of samples that are just 0 (so silence in practice), or do we call its process method but handing no samples at all (0 amount)? I assume we do the first one as that makes more sense.

Right, I thought the first one. The RubberBand is exactly with zero samples (with silence as you mentioned)


> Why do these dropouts occur?

The dropouts occur because RubberBand does not have sufficiently filled the inner buffers. I can't give you now for this some exactly clear answer, because it is more dependent on the RubberBand implementation and I'm not so deeply in it. Simply put, we haven't filled RubberBand with a sufficient amount of data for its inner processing and based on that, the RubberBand can't provide us so many samples on the output. It is a few times noted in discussions dependent on RubberBand, that it doesn't mean, that if some number of input frames are provided into RubberBand, we couldn't expect the same number of frames on the output, especially at the start of RubberBand processing. When we filled the RubberBand with zero samples, the inner implementation looks more stable and IMO when the zero frames are filled into the RubberBand, it can use for inner calculation with the real data and provide meaningful output much earlier. 

> What kind of dropouts is that strategy supposed to prevent?

I changed the order of this question because the answer can be based on the previous one. It prevents the incomplete buffer. If we want on the output 1024 frames but RubberBand provides only three-quarters of the required frames for example. With the zero frames prefilling, this situation does not occur (sometimes it can occur just only for mad pitch changes).

**Comment by davidchocholaty** [comment]
> So this version basically feeds rubberband silence until we have real audio data to serve?

Right, exactly.

**Comment by davidchocholaty** [comment]
> > When the last block of the input buffer is processed and the amount of frames that RubberBand requires is greater than the rest of the input buffer frames, only the rest of the input buffer frames are provided and not the amount, that RubberBand requires.
> 
> What advantage does this have?

It is not exactly an advantage but a disadvantage. I only described the difference between the "clean pull" model implementation and this one. The ideal situation would be, that for every `getSamplesRequired()` call we can provide exactly the requested amount of input frames, but with a fixed input buffer size it is not possible.

The more detailed answer was in the previous messages. If it is not clear to you still, please let me know.

**Comment by davidchocholaty** [comment]
> And my main question is still, why does rubberband not do all of this internally by default if it's better?

RubberBands recommends the "pull" model for real-time usage because it is designed for it as the authors mentioned. I think that I can't give you now a clear answer because it depends on the RubberBand implementation. IMO it really depends on the usage code implementation and how it works with the RubberBand, but on the other hand, if somebody asks the authors questions or somebody implements the effect with RubberBand usage, in most cases, it is related to the author's public [ladspa-lv2](https://hg.sr.ht/~breakfastquay/rubberband/browse/ladspa-lv2/RubberBandPitchShifter.cpp?rev=tip) implementation, here is the code part from the mentioned implementation for the "zero-prefilling" situation.

```
for (size_t c = 0; c < m_channels; ++c) {
    for (size_t i = 0; i < m_bufsize; ++i) {
        m_scratch[c][i] = 0.f;
    }
}

m_stretcher->process(m_scratch, m_reserve, false);
```

To really cleanly and understandably answer your question I think I would have to take a discussion with the authors if you like. Please give me feedback.

**Comment by Swiftb0y** [comment]
Ok thanks. I think you've answered all of my questions for now. It probably makes sense to summarize some of that info and put it in the code for posterity's sake.

**Comment by Swiftb0y** [comment]
Ok next question:
Why do we divide the input buffer up in blocks? Can't we just repeatedly call `RubberbandStretcher::getSamplesRequired` and give it as many samples as it wants until it wants more than we can give it?

**Comment by davidchocholaty** [comment]
> Ok thanks. I think you've answered all of my questions for now. It probably makes sense to summarize some of that info and put it in the code for posterity's sake.

No problem, great idea.

**Comment by daschuer** [comment]
This is an interesting Bit: 
https://breakfastquay.com/rubberband/code-doc/classRubberBand_1_1RubberBandStretcher.html#a3cd2ea056ec0eb94087f2fca2a41b080
I thing that is why our original approach equerry the group delay fails.  

I consider this: 
In case Rubberband has enough samples to to do the most difficult pitch value, it is in a settled state. Since the rate is always 1.0 in the effect case, it should exactly consume the same samples we take from it and the group delay can be constant. 

In case Rubberband requested more, it was not in this state initially and we need to fill more samples in. 
In case it provides more, we must not fetch them ore alternatively buffer them for later callbacks. 

Has this been implemented here?

File: src/effects/backends/builtin/pitchshifteffect.cpp [hunk 1]
```diff
@@ -89,28 +137,82 @@ void PitchShiftEffect::processChannel(
...
+        if (offsetFrames + blockFrames > framesPerBuffer) {
+            blockFrames = framesPerBuffer - offsetFrames;
+        }
+
+        SINT processedFrames = 0;
+
+        while (processedFrames < blockFrames) {
+            // RubberBand works with multichannel Sample FRAMES.
+            SINT requiredFrames = pState->m_pRubberBand->getSamplesRequired();
+            SINT toProcessFrames = math_min(blockFrames - processedFrames, requiredFrames);
```
Comment (daschuer) [inline src/effects/backends/builtin/pitchshifteffect.cpp:165]: What happens if requiredFrames > framesPerBuffer?

File: src/effects/backends/builtin/pitchshifteffect.cpp [hunk 2]
```diff
@@ -89,28 +137,82 @@ void PitchShiftEffect::processChannel(
...
+                    toProcessFrames,
+                    false);
+
+            processedFrames += toProcessFrames;
+
+            SINT availableFrames = pState->m_pRubberBand->available();
+            SINT writableFrames = pState->m_outputBuffer->getWriteSpace() / channelCount;
+
+            if (availableFrames > writableFrames) {
+                qDebug() << "PitchShiftEffect::processChannel: buffer is not large enough";
```
Comment (daschuer) [inline src/effects/backends/builtin/pitchshifteffect.cpp:184]: We must not qDebug() in the engine thread, because locking buffer allocation is involved.

File: src/effects/backends/builtin/pitchshifteffect.cpp [hunk 3]
```diff
@@ -89,28 +137,82 @@ void PitchShiftEffect::processChannel(
...
+                availableFrames = writableFrames;
+            }
+
+            SINT retrievedFrames = pState->m_pRubberBand->retrieve(
+                    pState->m_retrieveBuffer, availableFrames);
+            SampleUtil::interleaveBuffer(pState->m_interleavedBuffer,
```
Comment (daschuer) [inline src/effects/backends/builtin/pitchshifteffect.cpp:190]: I think it is more effective to do the interleaving outside the loop.

File: src/effects/backends/builtin/pitchshifteffect.cpp [hunk 4]
```diff
@@ -89,28 +137,82 @@ void PitchShiftEffect::processChannel(
...
+                    retrievedFrames);
+
+            pState->m_outputBuffer->write(pState->m_interleavedBuffer,
+                    retrievedFrames * channelCount);
+        }
+
+        SINT toReadFrames = pState->m_outputBuffer->getReadSpace() / channelCount;
+        SINT missingFrames = 0;
+
+        if (toReadFrames < blockFrames) {
```
Comment (daschuer) [inline src/effects/backends/builtin/pitchshifteffect.cpp:202]: Can we assert against this? I thing if this happens we need to pre-fill Rubberband more.

**Review by daschuer — CHANGES_REQUESTED** [review]
The ringbuffer should be probably moved to the input side. 

Currently we have 

Input -> Feed by chunk to Rubberband -> get chunks -> Ring, in case we receive more than we expect -> output.   

The output ring will always contain kPrefilledFrames - non const Rubberband group delay.


Alternative:  
Input -> into Ring -> Feed by chunk to Rubberband -> get chunks -> output. 

Start processing after input ring is filled > kPrefilledFrames and fill output with zeros in the meantime. 

This is more like the mode of the EngineBufferScaler, where the output controls the speed. And does not require to use CPU for settle Rubberband with zeros.

File: src/effects/backends/builtin/pitchshifteffect.cpp [hunk 5]
```diff
@@ -45,9 +45,52 @@ PitchShiftGroupState::PitchShiftGroupState(
...
+    //
+    // The X letter represents the real frame data.
+    //
+    // stream of input buffers:
+    //
+    // |00000000|00000000|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX| ...
+    //
+    // stream of output buffers:
+    //
+    //                   |00000000|00000000|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX| ...
```
Comment (daschuer) [inline src/effects/backends/builtin/pitchshifteffect.cpp:93]: What is the difference between 0 and X inside Rubberband? Isn't 0 just a special X? Probably the visual is just misleading.
Is each bar a callback? I think than you may visualize this: 

```
without zero initalization 
in:    |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX| ...
out:   |        |        |        |    XXXX|    XXXX|  XXXXXX|XXXXXXXX|XXXXXXXX| ...

6 x out broken, if we skip 5 cycles we have enough samples in the 6th,   

with two cycles zero initialization 
in:    |00000000|00000000|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX| ...
out:   |        |        |        |        |        |00000000|00000000|XXXXXXXX|XXXXXXXX| ...
 

with zero initalization during init 
init:  |00000000|00000000|00000000|00000000|00000000|
in:    |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX| ...
out:   |00000000|00000000|00000000|00000000|00000000|XXXXXXXX| ...

**Comment by davidchocholaty** [comment]
> This is an interesting Bit: https://breakfastquay.com/rubberband/code-doc/classRubberBand_1_1RubberBandStretcher.html#a3cd2ea056ec0eb94087f2fca2a41b080

Yeah, it looks cool for the R3 version. I have seen it once before and IMO it looks like they tried to make the similar approach as is used for the ladspa-lv2 implementation more user-friendly and for wider use cases.
 
> I thing that is why our original approach equerry the group delay fails.

Do you mean with the "original approach" the approach in this PR? Am I right?

> I consider this: In case Rubberband has enough samples to to do the most difficult pitch value, it is in a settled state.

Prefilling the RubberBand with silence samples serves this purpose. Based on the [ladspa-lv2](https://hg.sr.ht/~breakfastquay/rubberband/browse/ladspa-lv2/RubberBandPitchShifter.cpp?rev=tip#L284) implementation, the constant 8192 should be sufficient for all cases.

> Since the rate is always 1.0 in the effect case, it should exactly consume the same samples we take from it and the group delay can be constant.

That's what I tried to follow which was used in the mentioned [public authors implementation](https://hg.sr.ht/~breakfastquay/rubberband/browse/ladspa-lv2/RubberBandPitchShifter.cpp?rev=tip#L462) too. The group delay constant is the same as the number of prefilled silence samples. Anyway, for my usage, it works only for the first Pitch shift effect use. After some dynamic Pitch changes, the RubberBand output samples do not match the input samples (if we still calculate with the constant group delay), but it can be caused by some hidden bug now.



> In case Rubberband requested more, it was not in this state initially and we need to fill more samples in.

Sounds reasonable, I agree. Anyway, for the usage of 8192 prefilled silence samples, for some weird dynamic Pitch changes, the RubberBand can have a problem filling the whole output buffer. Probably, it can be solved by prefilling more silence samples, however, it looks to me like it just increases an effect delay a lot.

> In case it provides more, we must not fetch them ore alternatively buffer them for later callbacks.
> 
> Has this been implemented here?

Yeah, for that problem as you mentioned later serves the output ring buffer, where the unwritten samples into the output buffer are stored for the next callback.

File: src/effects/backends/builtin/pitchshifteffect.cpp (hunk 5 above)
Comment (daschuer) [reply src/effects/backends/builtin/pitchshifteffect.cpp:93]: My idea is that we can do the same wit a start delay and real samples: 

```
without zero initalization 
in:    |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX| ...
out:   |        |        |        |    XXXX|    XXXX|  XXXXXX|XXXXXXXX|XXXXXXXX| ...

6 x out broken, if we skip 5 cycles we have enough samples in the 6th,  
Fill output with 0 during the first 5 cycles  

in:    |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX| ...
out:   |00000000|00000000|00000000|00000000|00000000|XXXXXXXX|XXXXXXXX|XXXXXXXX| ...

since de group delay is not const put some X into a ring and comsume it whenever Rubberband request it via getSamplesRequired() 
 
in:    |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX| ...
out:   |00000000|00000000|00000000|00000000|00000000|00000000|XXXXXXXX|XXXXXXXX| ...

```

File: src/effects/backends/builtin/pitchshifteffect.cpp [hunk 6]
```diff
@@ -160,6 +203,41 @@ void PitchShiftEffect::processChannel(
...
+            // like the "pull" implementation as known for RubberBand
+            // implementation, anyway, it is not for example
+            // for the following reasons:
+            //
+            // 1. Prefilling with zero samples to avoid dropouts.
+            // 2. When the last block of the input buffer is processed
+            //    and the amount of frames that RubberBand requires is greater
+            //    than the rest of the input buffer frames, only the rest
+            //    of the input buffer frames are provided and not the amount,
+            //    that RubberBand requires.
```
Comment (daschuer) [inline src/effects/backends/builtin/pitchshifteffect.cpp:216]: Can RubberBand even use the rest, or is something bad happening. 

I expect the Rubberband users windowing functions. With 50 % overlap or such. If you now pass only a few samples the window is not full. What will happen? Rubberband fills is with old samples, CPU is increase / Rubberband does nothing, than there is no pint to even provide the samples / Rubberband pads the window with zeros which may create an audible glitch. (I have not checked so I might be wrong).

From this point of view a traditional pull model with an input buffer creating a fixed group delay seems to be more robust. This way we can always provide the expected samples for the desired output with no exceptions.

**Comment by davidchocholaty** [comment]
> > > I thing that is why our original approach equerry the group delay fails.
> 
> > Do you mean with the "original approach" the approach in this PR? Am I right?
> 
> No, I mean the merged PR where the queried group delay from Rubberband is reported wrong.


Oh, right. We agreed to move the group delay handling exactly for the Pitch shift effect into separated PR, so, the wrong delay reporting was removed from the merged PR. It uses the default set value, which is 0 (without delay). But you are right, based on some missing samples due to the RubberBand was not prefilled with silence, the reported value was wrong.

**Comment by daschuer** [comment]
> Based on the [ladspa-lv2](https://hg.sr.ht/~breakfastquay/rubberband/browse/ladspa-lv2/RubberBandPitchShifter.cpp?rev=tip#L284) implementation, the constant 8192 should be sufficient for all cases.

Cool, does it depends on the sample rate? I think we can find out it easily if we watch the ring buffer load in this implementation.

> but it can be caused by some hidden bug now.

I expect they will vanish if we move this to a pull model from a ring buffer with size 8192. 

> ... it looks to me like it just increases an effect delay a lot.

Since to much group delay makes this effect hard to use and after this is merged we have always the worst case group delay. 
We need to find out the pitch range that works with a reasonable group delay and the pitch limits accordingly.

**Comment by davidchocholaty** [comment]
> Cool, does it depends on the sample rate? I think we can find out it easily if we watch the ring buffer load in this implementation.

Yeah, you're right. If I will base it on the R3 finer implementation (which should be used after introducing the R3 version), it works with the following calculation, if I found it right:

- [getPreferredStartPad](https://github.com/breakfastquay/rubberband/blob/dca4c9683401af5495373e4e15b46af0bfbe83bb/src/finer/R3Stretcher.cpp#L412)
- [sample rate usage](https://github.com/breakfastquay/rubberband/blob/567e8636cdcc5407b8b9c15ed2a34b52850b560b/src/finer/Guide.h#L113)

File: src/effects/backends/builtin/pitchshifteffect.cpp (hunk 6 above)
Comment (daschuer) [reply src/effects/backends/builtin/pitchshifteffect.cpp:212]: rubberband actually skips the process call if we have provided not enough data: 
https://github.com/breakfastquay/rubberband/blob/10661f40c60038cef55c1a2d8e417dbca4a17616/src/faster/StretcherProcess.cpp#L335

**Comment by davidchocholaty** [comment]
> I expect they will vanish if we move this to a pull model from a ring buffer with size 8192.

Oh, I think that I take it now all together. Based on your proposals, using the ring buffer on the input side with optional size ensures, that we can work with RubberBand exactly as in the "pull" model instead of this hybrid model between "pull" and "push". If I am right with that, sounds really awesome.

**Comment by davidchocholaty** [comment]
> Since to much group delay makes this effect hard to use and after this is merged we have always the worst case group delay.
> We need to find out the pitch range that works with a reasonable group delay and the pitch limits accordingly.

Sounds reasonable, agree.

**Comment by daschuer** [comment]
> I am right with that, sounds really awesome.

Yes, that was the idea.

**Comment by davidchocholaty** [comment]
> > I am right with that, sounds really awesome.
> 
> Yes, that was the idea.

Perfect, thank you.

**Comment by Swiftb0y** [comment]
@davidchocholaty also discussed an alternative solution to the one implemented in this PR:

We keep the push model as currently present within main and keep a buffer of delayed audio data and whenever we would encounter dropouts or possibly incomplete buffers, we just use the delayed input signal instead.

**Comment by daschuer** [comment]
How does the delayed samples fit to the current? Do you plan to fade over? Even than we will have an audible glitch. 
The basic idea in this PR with the dynamic puffer for a constant group delay is probably a better solution, except that we have always the maximum delay. Do we have an idea in  which range the delay changes when changing the pitch?

**Comment by Swiftb0y** [comment]
> How does the delayed samples fit to the current? Do you plan to fade over? Even than we will have an audible glitch.

Yes, but there will always be an audible glitch. Either we leave the buffer empty, or we fill with roughly matching data (the unstretched one). Either way, we currently have no good plan on how to handle the events where latency increases or decreases.

**Comment by Swiftb0y** [comment]
Even with the prefilling solution, we'll have to plan for the eventual case that we can't fill the output buffer because rubberband won't give us any samples.

**Comment by Swiftb0y** [comment]
> How does the delayed samples fit to the current?

We use rubberbands `getLatency`/`getStartDelay` to get the correct offset.

**Comment by Swiftb0y** [comment]
We have a basic tradeoff to make here. We can either pre-cache a bunch resulting in huge delays/latency, or we can try to minimize the delay and try to hide any dropouts.

**Comment by daschuer** [comment]
For a decision, a graph group delay vs pitch change would be nice. This can be done with the metronome effect chained with the pitch shift effect, recorded with dry/whet at 3 o clock. 
Than you can measure distance if the lower volume click to the higher one using audacity.

(The result is perfectly for your blog post as well) 

Hiding a dropout is quite hard because we need to process samples, we do not have if the buffer is running low. So we may add the reverse of the last buffer or such tricks. And then fade into the new samples. Just fade to zero and fade in from zero once we have new samples may also be OK. In addition the track will fall out beat sync creating a double base effect.

**Comment by davidchocholaty** [comment]
I totally agree, the comparison shows us which approach is better and which we should prefer. There is just a little bug I found out and should be fixed or "masked" for testing for now. The RubberBand works normally, anyway the value passed to it in the [construction of the RubberBand stretcher](https://github.com/mixxxdj/mixxx/blob/669d4519c2a54e9855174b2d01ad8e379e77ba7d/src/effects/backends/builtin/pitchshifteffect.cpp#L27) is invalid, specifically the `engineParameters.sampleRate()` value is invalid due to [this TODO](https://github.com/mixxxdj/mixxx/blob/d8826cab4d9155ab1ffd5dcc77b603b696292017/src/engine/effects/engineeffect.cpp#L33) in the EngineEffect which stores just the maximum possible value. IMO based on that we can maybe make the input ring buffer just smaller due to inner RubberBand values have changed too.

I don't want this fix just brake the project in this part, so I would like to propose just creating the stretcher if it is not in the `PitchShiftEffect::processChannel` and fixing the bug later. Do you agree? I would like to create a GitHub issue for this problem if it is not.

**Comment by davidchocholaty** [comment]
> For a decision, a graph group delay vs pitch change would be nice. This can be done with the metronome effect chained with the pitch shift effect, recorded with dry/whet at 3 o clock. Than you can measure distance if the lower volume click to the higher one using audacity.
> 
> (The result is perfectly for your blog post as well)
> 
> Hiding a dropout is quite hard because we need to process samples, we do not have if the buffer is running low. So we may add the reverse of the last buffer or such tricks. And then fade into the new samples. Just fade to zero and fade in from zero once we have new samples may also be OK. In addition the track will fall out beat sync creating a double base effect.

Thank you very much for the advice. I would like just to ask how about it is possible for the metronome effect to play separately from the Pitch shift effect output for measurement when they are in the same effect chain. I suppose, that you probably propose something like playing a track with a recorded metronome, then turn on the Pitch shift effect with mentioned wet/dry setting and record the final output via Mixxx to have tracks for beats comparing in DAW. I have some tracks with clicks available, so I can use that. I will publicise the measuring results asap.

**Comment by davidchocholaty** [comment]
I have some results from the measuring. The measuring was provided by following the procedure below:

1. Play track with a recorded metronome (click sound).
2. Turn the Pitch shift effect on set in the middle position (zero pitch).
3. After a few bars change the Pitch value continuously to +1 octave value, keep it for a few bars and continuously get back to the zero pitch and again keep it for a few bars.
4. Then repeat a similar situation but just until the -1 octave Pitch value is reached and get back into the zero pitch setting.

All this workflow was repeated to the merged "push" implementation and the new "pull" implementation with different sizes of the input ring buffer (8192 frames or 4096 frames).

Every recorded track was checked by using DAW and the group delay frames were calculated using the samples grid in the DAW. All values were plotted into the following two graphs.

-------------------------------

My parameters:
- Sample Rate: 44100 Hz
- Frames per buffer: 1024

Comparison for the specific Pitch value settings (-1 octave, 0 pitch, +1 octave):
-------------------------------
comparision


Comparison for the zero pitch but the previous set Pitch value was different:
-------------------------------
comparision_zero_pitch

**Comment by daschuer** [comment]
What is your conclusion? 

It looks like the pull model is the way to go.

Am I right that Rubberband has a kind of buffer leveling internally. To dill that down I have decorated the mixer pitch shifter with some qDebug() statements: 
https://github.com/daschuer/mixxx/tree/rb_buffer.

Interesting for us is the first set of messages after reset() for differen pitch settings.
There seems to be a dependency higher pitch = bigger buffer.

How about setting our caching buffer to the max value selected by the range knob? Than we can promise: no droputs when using "pitch" but notable delay adjustments when touching range.

**Comment by daschuer** [comment]
While testing I have discovered this old bug/design issue: https://github.com/mixxxdj/mixxx/issues/10833

**Comment by davidchocholaty** [comment]
> What is your conclusion?
> 
> It looks like the pull model is the way to go.

Agree. It looks like it can work but depends really on the input ring buffer size.

From the second graph, the group delay after dynamic pitch changes is more stable than for the "push" model. Anyway, it is not apparent from the graph, that the group delay is not constant in fact, there are little changes but not so huge that can be visible. 

For the 4096 input ring buffer size, it was about 3 frames difference, for 8192 about 30 frames difference. For now, I haven't found the reason why, but it is true, that if the group delay value is set properly as constant, for the first use for the Pitch middle position (pitch is unchanged), the wet and dry signal overlap. However, after dynamic changes the wet and dry signals are a little bit shifted to each other.

> Am I right that Rubberband has a kind of buffer leveling internally. To drill that down I have decorated the mixer pitch shifter with some qDebug() statements: https://github.com/daschuer/mixxx/tree/rb_buffer.
> 
> Interesting for us is the first set of messages after reset() for differen pitch settings. There seems to be a dependency higher pitch = bigger buffer.

Thank you for your research and the link. I tried it yourself too and it probably looks like you're right.

> How about setting our caching buffer to the max value selected by the range knob? Than we can promise: no droputs when using "pitch" but notable delay adjustments when touching range.

Yeah, makes sense. Anyway, we should take the sample rate into account as well due to when the sample rate value which is passed into the RubberBand when creating the stretcher is different, the required frame amount is also different.

> We may use two or three fixed delay buffer sizes and use them according to the range knob.

Cool idea. I suppose you have in mind the input ring buffer under "delay buffer", just to be sure. I'm just thinking about the production realization of this idea.

If the ring buffer size will be just the biggest one, we can start processing when the number of frames needed for the range setting processing was filled into it (mentioned fixed delay buffer size). Anyway, I think, that it can occur the situation, that we will start processing and produce output for some range setting, then the range setting can dramatically change and in this situation, we started processing but we don't have to still enough samples in the ring buffer, that we can process data with the new setting, so the dropout has occurred.

Correct me if I am wrong.

**Comment by daschuer** [comment]
> we started processing but we don't have to still enough samples in the ring buffer ... 

yes that is an issue. If we adjust the range. In that case we can just fade to zero, wait the additional delay and fade from zero. This removes at least the click noise. The user will learn quickly that touching the range knob during playing out mater is not a good idea.  The first issue of that type is the first callback. The delay increase from zero to something.

I have not jet fully understand why you prefer the buffer at the output. I consider an input buffer and the pure pull model simpler. It would be a pity if Rubberband increases its internal buffer by this model. In that case the output buffering seems the best solution. We just should be clear to avoid unnecessary complexity and add some notes for later reference.

**Comment by davidchocholaty** [comment]
> yes that is an issue. If we adjust the range. In that case we can just fade to zero, wait the additional delay and fade from zero. This removes at least the click noise.

Agree, it could be a solution.

> The user will learn quickly that touching the range knob during playing out mater is not a good idea.


Probably yes. I just remember that when we designed the Pitch shift effect extended option the way we use now (with the Range knob), it looks cool to me, that we can just set the Pitch knob at the maximum or minimum value and turn with the Range knob to pitch only up or only down like on mentioned Pioneer mixer. But agree, that we just have to make here a tradeoff to decrease the effect latency.

> I have not jet fully understand why you prefer the buffer at the output.

I do not prefer the buffer at the output since your proposal with move in into the input and which way it is implemented in the new version of this PR. Sorry for the misunderstanding, if I wrote something not so clearly.

> I consider an input buffer and the pure pull model simpler.

Agree.
